packages/styled-system/src/css-var.ts

```diff
@@ -0,0 +1,186 @@
+import { Dict, isCssVar, isObject, pick } from "@chakra-ui/utils"
+import analyzeBreakpoints from "./breakpoints"
+import type { CSSMap, WithCSSVar } from "./types"
+
+const separator = ">"
+
+const replace = (value: string, str = ".") => value.replace(/>/g, str)
+
+export const toVar = (value: string) => `var(${value})`
+
+export const tokens = [
+  "colors",
+  "borders",
+  "borderWidths",
+  "borderStyles",
+  "fonts",
+  "fontSizes",
+  "fontWeights",
+  "letterSpacings",
+  "lineHeights",
+  "radii",
+  "space",
+  "shadows",
+  "sizes",
+  "zIndices",
+  "transitions",
+  "transition.duration",
+  "transition.property",
+  "transition.easing",
+] as const
+
+export type ThemeScale = typeof tokens[number]
+
+function extractTokens(theme: Dict) {
+  const _tokens = (tokens as unknown) as string[]
+  return pick(theme, _tokens)
+}
+
+const getKeys = (key: string) => ({
+  varKey: `--${replace(key, "-")}`,
+  mapKey: replace(key, "."),
+  negVarKey: `--${key.replace(">-", "-")}`,
+})
+
+interface AssignOptions {
+  cssVars: Dict
+  cssMap: CSSMap
+  key: any
+  value: any
+  negate?: boolean
+}
+
+function assign(options: AssignOptions) {
+  const { cssVars, cssMap, key, value, negate: minus } = options
+
+  const _value = minus ? negate(value) : value
+  const { varKey, mapKey, negVarKey } = getKeys(key)
+
+  if (!minus) {
+    cssVars[varKey] = _value
+  }
+
+  const _varKey = minus ? negVarKey : varKey
+  const minusVal = isCssVar(value)
+    ? `calc(${value} * -1)`
+    : `calc(${toVar(_varKey)} * -1)`
+
+  const varRef = minus ? minusVal : toVar(varKey)
+  cssMap[mapKey] = {
+    var: _varKey,
+    value: isCssVar(value) ? value : _value,
+    varRef,
+  }
+}
+
+function negate(value?: string) {
+  if (!value) return
+  const num = parseFloat(value)
+  const unit = String(value).replace(String(num), "")
+  return `-${num}${unit}`
+}
+
+function omitVars(theme: Dict) {
+  if ("__cssMap" in theme) {
+    delete theme.__cssMap
+    delete theme.__cssVars
+    delete theme.__breakpoints
+  }
+}
+
+/**
+ * The CSS transform order following the upcoming spec from CSSWG
+ * translate => rotate => scale => skew
+ * @see https://drafts.csswg.org/css-transforms-2/#ctm
+ * @see https://www.stefanjudis.com/blog/order-in-css-transformation-transform-functions-vs-individual-transforms/
+ */
+const transformTemplate = [
+  "rotate(var(--rotate, 0))",
+  "scaleX(var(--scale-x, 1))",
+  "scaleY(var(--scale-y, 1))",
+  "skewX(var(--skew-x, 0))",
+  "skewY(var(--skew-y, 0))",
+]
+
+export function getTransformTemplate(type?: "gpu") {
+  const withGpu = "translate3d(var(--translate-x, 0), var(--translate-y, 0), 0)"
+  const basic = [
+    "translateX(var(--translate-x, 0))",
+    "translateY(var(--translate-y, 0))",
+  ].join(" ")
+  const transform = type === "gpu" ? withGpu : basic
+  return [transform, ...transformTemplate].join(" ")
+}
```

Comment:
This function does two things at once. I think we can add more readability by splitting it up in two separate functions:

```suggestion
export function getTransformTemplate() {
  return [ 
    "translateX(var(--translate-x, 0))",
    "translateY(var(--translate-y, 0))",
    ...transformTemplate
  ].join(" ")
}

export function getTransformGpuTemplate(type?: "gpu") {
 return [
   "translate3d(var(--translate-x, 0), var(--translate-y, 0), 0)",
    ...transformTemplate
  ].join(" ")
}
```